File: PE-diagram/src/main/java/nl/adaptivity/process/diagram/DrawableStartNode.java
/*
 * Copyright (c) 2017.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.diagram;
import nl.adaptivity.diagram.*;
import nl.adaptivity.process.clientProcessModel.ClientStartNode;
import nl.adaptivity.process.processModel.IXmlDefineType;
import nl.adaptivity.process.processModel.IXmlResultType;
import nl.adaptivity.process.processModel.StartNode;
import nl.adaptivity.process.util.Identified;
import nl.adaptivity.xml.XmlException;
import nl.adaptivity.xml.XmlReader;
import nl.adaptivity.xml.XmlUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;

import static nl.adaptivity.process.diagram.RootDrawableProcessModel.*;



public class DrawableStartNode extends ClientStartNode<DrawableProcessNode, DrawableProcessModel> implements DrawableProcessNode{

  public static class Builder extends ClientStartNode.Builder<DrawableProcessNode, DrawableProcessModel> implements DrawableProcessNode.Builder {

    public Builder() { }

    public Builder(final boolean compat) {
      super(compat);
    }

    public Builder(@Nullable final Identified successor, @Nullable final String id, @Nullable final String label, final double x, final double y, @NotNull final Collection<? extends IXmlDefineType> defines, @NotNull final Collection<? extends IXmlResultType> results) {
      super(successor, id, label, x, y, defines, results);
    }

    public Builder(@NotNull final StartNode<?, ?> node) {
      super(node);
    }

    @NotNull
    @Override
    public DrawableStartNode build(@NotNull final DrawableProcessModel newOwner) {
      return new DrawableStartNode(this, newOwner);
    }
  }

  private static final double REFERENCE_OFFSET_X = STARTNODERADIUS + (STROKEWIDTH / 2);
  private static final double REFERENCE_OFFSET_Y = STARTNODERADIUS + (STROKEWIDTH / 2);
  public static final String IDBASE = "start";
  private int mState = STATE_DEFAULT;

  public DrawableStartNode(final DrawableProcessModel ownerModel) {
    this(ownerModel, false);
  }

  public DrawableStartNode(final DrawableProcessModel ownerModel, final boolean compat) {
    super(ownerModel, compat);
  }

  public DrawableStartNode(final DrawableProcessModel ownerModel, String id, final boolean compat) {
    super(ownerModel, id, compat);
  }

  public DrawableStartNode(DrawableStartNode orig, final boolean compat) {
    super(orig, compat);
    mState = orig.mState;
  }

  public DrawableStartNode(@NotNull final StartNode.Builder<?, ?> builder, @Nullable final DrawableProcessModel newOwnerModel) {
    super(builder, newOwnerModel);
  }

  @NotNull
  @Override
  public Builder builder() {
    return new Builder(this);
  }

  @Override
  public DrawableStartNode clone() {
    if (getClass()==DrawableStartNode.class) {
      return new DrawableStartNode(this, isCompat());
    }
    throw new RuntimeException(new CloneNotSupportedException());
  }

  @NotNull
  public static Builder deserialize(@NotNull final XmlReader in) throws XmlException {
    return XmlUtil.deserializeHelper(new Builder(true), in);
  }

  @Deprecated
  @NotNull
  public static DrawableStartNode deserialize(final DrawableProcessModel ownerModel, @NotNull final XmlReader in) throws XmlException {
    return XmlUtil.<DrawableStartNode.Builder>deserializeHelper(new DrawableStartNode.Builder(true), in).build(ownerModel);
  }

  @Override
  public Rectangle getBounds() {
    return new Rectangle(getX()-REFERENCE_OFFSET_X, getY()-REFERENCE_OFFSET_Y, STARTNODERADIUS * 2 + STROKEWIDTH, STARTNODERADIUS * 2 + STROKEWIDTH);
  }

  @Override
  public void translate(double dX, double dY) {
    setX(getX() + dX);
    setY(getY() + dY);
  }

  @Override
  public void setPos(double left, double top) {
    setX(left+REFERENCE_OFFSET_X);
    setY(left+REFERENCE_OFFSET_Y);
  }

  @Override
  public boolean isWithinBounds(final double x, final double y) {
    final double realradius= STARTNODERADIUS + (STROKEWIDTH / 2);
    return ((Math.abs(x-getX())<=realradius) && (Math.abs(y-getY())<=realradius));
  }

  @Override
  public Drawable getItemAt(double x, double y) {
    return isWithinBounds(x, y) ? this : null;
  }

  @Override
  public int getState() {
    return mState ;
  }

  @Override
  public void setState(int state) {
    mState = state;
  }

  @Override
  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> void draw(Canvas<S, PEN_T, PATH_T> canvas, Rectangle clipBounds) {
    if (hasPos()) {
      final double realradius= STARTNODERADIUS + (STROKEWIDTH / 2);
      PEN_T fillPen = canvas.getTheme().getPen(ProcessThemeItems.LINEBG, mState & ~STATE_TOUCHED);

      if ((mState&STATE_TOUCHED)!=0) {
        PEN_T touchedPen = canvas.getTheme().getPen(ProcessThemeItems.LINE, STATE_TOUCHED);
        canvas.drawCircle(realradius, realradius, STARTNODERADIUS, touchedPen);
      }

      canvas.drawFilledCircle(realradius, realradius, realradius, fillPen);
    }
  }

  @Override
  public String getIdBase() {
    return IDBASE;
  }

  @Override
  public <S extends DrawingStrategy<S, PEN_T, PATH_T>, PEN_T extends Pen<PEN_T>, PATH_T extends DiagramPath<PATH_T>> void drawLabel(Canvas<S, PEN_T, PATH_T> canvas, Rectangle clipBounds, double left, double top) {
    DrawableUtil.defaultDrawLabel(this, canvas, clipBounds, left, top);
  }

  public static DrawableStartNode from(StartNode<?, ?> n, final boolean compat) {
    DrawableStartNode result = new DrawableStartNode((DrawableProcessModel) null, compat);
    copyProcessNodeAttrs(n, result);
    result.setDefines(n.getDefines());
    return result;
  }

}

File: PE-diagram/src/commonMain/kotlin/nl.adaptivity.process/diagram/DrawableStartNode.kt
/*
 * Copyright (c) 2017.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.diagram

import nl.adaptivity.diagram.*
import nl.adaptivity.process.diagram.RootDrawableProcessModel.Companion.STARTNODERADIUS
import nl.adaptivity.process.diagram.RootDrawableProcessModel.Companion.STROKEWIDTH
import nl.adaptivity.process.processModel.*
import nl.adaptivity.process.util.Identified
import nl.adaptivity.xmlutil.XmlReader

interface IDrawableStartNode : IDrawableProcessNode {
    override val leftExtent get() = REFERENCE_OFFSET_X
    override val rightExtent get() = STARTNODERADIUS * 2 + STROKEWIDTH - REFERENCE_OFFSET_X
    override val topExtent get() = REFERENCE_OFFSET_Y
    override val bottomExtent get() = STARTNODERADIUS * 2 + STROKEWIDTH - REFERENCE_OFFSET_Y

    override val maxPredecessorCount: Int get() = 0

    override fun <S : DrawingStrategy<S, PEN_T, PATH_T>, PEN_T : Pen<PEN_T>, PATH_T : DiagramPath<PATH_T>> draw(canvas: Canvas<S, PEN_T, PATH_T>,
                                                                                                                clipBounds: Rectangle?) {
        if (hasPos()) {
            val realradius = STARTNODERADIUS + STROKEWIDTH / 2
            val fillPen = canvas.theme.getPen(ProcessThemeItems.LINEBG, state and Drawable.STATE_TOUCHED.inv())

            if (state and Drawable.STATE_TOUCHED != 0) {
                val touchedPen = canvas.theme.getPen(ProcessThemeItems.LINE, Drawable.STATE_TOUCHED)
                canvas.drawCircle(realradius, realradius, STARTNODERADIUS, touchedPen)
            }

            canvas.drawFilledCircle(realradius, realradius, realradius, fillPen)
        }
    }

    companion object {
        const val REFERENCE_OFFSET_X = STARTNODERADIUS + STROKEWIDTH / 2
        const val REFERENCE_OFFSET_Y = STARTNODERADIUS + STROKEWIDTH / 2
    }

}

class DrawableStartNode(builder: StartNode.Builder,
                        buildHelper: ProcessModel.BuildHelper<*,*,*,*>) :
    StartNodeBase<DrawableProcessNode, DrawableProcessModel?>(builder, buildHelper),
    DrawableProcessNode {

    class Builder : StartNodeBase.Builder, DrawableProcessNode.Builder<DrawableStartNode>, IDrawableStartNode {

        override val _delegate: DrawableProcessNode.Builder.Delegate

        constructor() : this(id = null)

        constructor(id: String? = null,
                    successor: Identified? = null,
                    label: String? = null,
                    defines: Collection<IXmlDefineType> = emptyList(),
                    results: Collection<IXmlResultType> = emptyList(),
                    x: Double = Double.NaN,
                    y: Double = Double.NaN,
                    state: DrawableState = Drawable.STATE_DEFAULT,
                    isCompat: Boolean = false,
                    isMultiInstance: Boolean = false) : super(id, successor, label, defines, results, x, y,
                                                              isMultiInstance) {
            _delegate = DrawableProcessNode.Builder.Delegate(state, isCompat)
        }

        constructor(node: StartNode) : super(node) {
            _delegate = DrawableProcessNode.Builder.Delegate(node)
        }

        override fun copy(): Builder {
            return Builder(id, successor?.identifier, label, defines, results, x, y, state, isCompat, isMultiInstance)
        }

    }

    override val _delegate = DrawableProcessNode.Delegate(builder)

    override val idBase: String
        get() = IDBASE

    override val maxSuccessorCount: Int
        get() = if (isCompat) Int.MAX_VALUE else 1

    override val maxPredecessorCount get() = super.maxPredecessorCount

    override fun builder(): Builder {
        return Builder(this)
    }

    companion object {

        const val REFERENCE_OFFSET_X = STARTNODERADIUS + STROKEWIDTH / 2
        const val REFERENCE_OFFSET_Y = STARTNODERADIUS + STROKEWIDTH / 2
        const val IDBASE = "start"

        @kotlin.jvm.JvmStatic
        fun from(n: StartNode, compat: Boolean = false) = Builder(n).apply { this.isCompat = compat }.build()
    }

}
